### PR TITLE
Add support for digests

### PR DESCRIPTION
**Pull request by tatry**

### CODE REVIEW

File: lib/btf.c [hunk 1]
```diff
@@ -10,80 +10,85 @@
...
 {
-    type_id = follow_typedefs(btf, type_id);
+    type_id = follow_types(btf, type_id);
     if (type_id == 0)
         return NULL;
     return btf__type_by_id(btf, type_id);
 }
 
-uint32_t psabtf_get_type_id_by_name(struct btf * btf, const char * name)
+static uint32_t psabtf_get_map_type_id_by_name(struct btf * btf, const char * name)
```
Comment (kmateuszssak) [inline lib/btf.c:64]: `name` is no longer used?

Comment (kmateuszssak) [reply lib/btf.c:64]: Ok, I see.

**Review by kmateuszssak — COMMENTED** [review]
For now only two questions.

File: CLI/digest.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        if (psabpf_digest_get_field_type(field) == DIGEST_FIELD_TYPE_STRUCT_END)
+            return NO_ERROR;
+
+        if (psabpf_digest_get_field_type(field) != DIGEST_FIELD_TYPE_DATA)
+            continue;
+
+        const char *encoded_data = convert_data_to_hexstr(psabpf_digest_get_field_data(field),
+                                                          psabpf_digest_get_field_data_len(field));
+        const char *field_name = psabpf_digest_get_field_name(field);
+        if (encoded_data == NULL) {
```
[the remark this answers is not in the document]
Comment (tatry) [reply CLI/digest.c:74]: Yes, it should be.

File: CLI/common.c [hunk 3]
```diff
@@ -185,3 +187,45 @@ int translate_data_to_bytes(const char *data, void *ctx, enum destination_ctx_ty
     /* Last chance: parse as number */
     return convert_number_to_bytes(data, ctx, ctx_type);
 }
+
+char * convert_data_to_hexstr(const void *data, size_t len)
```
[the remark this answers is not in the document]
Comment (kmateuszssak) [reply CLI/common.c]: ok

File: CLI/digest.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    psabpf_digest_context_init(&ctx);
+
+    if (parse_pipeline_id(&argc, &argv, &psabpf_ctx) != NO_ERROR)
+        goto clean_up_psabpf;
+
+    if (parse_digest(&argc, &argv, &psabpf_ctx, &ctx, &digest_instance_name) != NO_ERROR)
+        goto clean_up_psabpf;
+
+    if (argc > 0) {
+        fprintf(stderr, "%s: unused argument\n", *argv);
```
Comment (kmateuszssak) [inline CLI/digest.c:104]: Maybe it would be better to show only a warning ?

File: CLI/common.c [hunk 5]
```diff
@@ -185,3 +187,45 @@ int translate_data_to_bytes(const char *data, void *ctx, enum destination_ctx_ty
...
+{
+    if (data == NULL)
+        return NULL;
+
+    size_t buff_len = len * 2 + 2 + 1; /* 2 characters per byte, prefix, null terminator */
+    char *buff = malloc(buff_len);
+    if (buff == NULL)
+        return NULL;
+
+    memset(buff, 0, buff_len);
```
Comment (kmateuszssak) [inline CLI/common.c:201]: Use calloc instad?

**Review by kmateuszssak — COMMENTED** [review]
Two more.

File: CLI/digest.c (hunk 4 above)
Comment (tatry) [reply CLI/digest.c:104]: I think it is better to make error for unused arguments:
- In [other places in our code](https://github.com/P4-Research/psabpf/blob/master/CLI/table.c#L256-L259) is done in this way.
- It is clear message to user that he made a mistake (`psabpf-ctl digest get` can obtain data from single digest instance, but what if user wants read two different Digests?)

File: CLI/common.c (hunk 5 above)
Comment (tatry) [reply CLI/common.c]: Done

**Review by osinstom — APPROVED** [review]
LGTM overall, would be nice to see a short demo tomorrow.

**Review by kmateuszssak — APPROVED** [review]

**Pull request merged**